### PR TITLE
More i2cdevices

### PR DESCRIPTION
**Pull request by lgywata**
Two more drivers using I2CDevice: PulseLightLRF and LightWareI2C. We do not have the hw to test these. @lucasdemarchi

### CODE REVIEW

File: libraries/AP_RangeFinder/RangeFinder.cpp [hunk 1]
```diff
@@ -473,15 +473,15 @@ void RangeFinder::add_backend(AP_RangeFinder_Backend *driver)
 
     drivers[num_instances++] = driver;
 }
-    
+
 /*
-  detect if an instance of a rangefinder is connected. 
+  detect if an instance of a rangefinder is connected.
```
Comment (lucasdemarchi) [inline libraries/AP_RangeFinder/RangeFinder.cpp]: unrelated changes above... add them in another commit please.

File: libraries/AP_RangeFinder/RangeFinder.cpp [hunk 2]
```diff
@@ -499,12 +499,9 @@ void RangeFinder::detect_instance(uint8_t instance)
...
     }
     if (type == RangeFinder_TYPE_LWI2C) {
-        if (AP_RangeFinder_LightWareI2C::detect(*this, instance)) {
-            state[instance].instance = instance;
-            drivers[instance] = new AP_RangeFinder_LightWareI2C(*this, instance, state[instance]);
-            return;
-        }
-    } 
+        add_backend(AP_RangeFinder_LightWareI2C::detect(*this, instance, state[instance],
+            hal.i2c_mgr->get_device(0, _address[instance])));
```
Comment (lucasdemarchi) [inline libraries/AP_RangeFinder/RangeFinder.cpp]: maybe check the _address[instance] here and don't create in this case

File: libraries/AP_RangeFinder/AP_RangeFinder_LightWareI2C.h [hunk 3]
```diff
@@ -3,21 +3,23 @@
...
 {
 
 public:
     // constructor
-    AP_RangeFinder_LightWareI2C(RangeFinder &ranger, uint8_t instance, RangeFinder::RangeFinder_State &_state);
+    AP_RangeFinder_LightWareI2C(RangeFinder &ranger, uint8_t instance, RangeFinder::RangeFinder_State &_state, AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev);
 
     // static detection function
-    static bool detect(RangeFinder &ranger, uint8_t instance);
+    static AP_RangeFinder_Backend *detect(RangeFinder &ranger, uint8_t instance, RangeFinder::RangeFinder_State &_state, AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev);
```
Comment (lucasdemarchi) [inline libraries/AP_RangeFinder/AP_RangeFinder_LightWareI2C.h]: In this case you need to make the constructor private.

File: libraries/AP_RangeFinder/AP_RangeFinder_LightWareI2C.cpp [hunk 4]
```diff
@@ -34,39 +36,45 @@ AP_RangeFinder_LightWareI2C::AP_RangeFinder_LightWareI2C(RangeFinder &_ranger, u
...
+    
+    if (ranger._address[state.instance] == 0) {
         return false;
     }
 
+    uint8_t buff[2];
     // read the high and low byte distance registers
-    if (hal.i2c->read(ranger._address[state.instance], 2, &buff[0]) != 0) {
-        i2c_sem->give();
+    if (!_dev->read_registers(ranger._address[state.instance], buff, sizeof(buff))) {
```
Comment (lucasdemarchi) [inline libraries/AP_RangeFinder/AP_RangeFinder_LightWareI2C.cpp]: wow... this is wrong. You already have the address of the device inside _dev, you shouldn't pass it here. This is rather:

```
if (!_dev->read_registers(2, buff, sizeof(buff))) {
```

Comment (lucasdemarchi) [inline libraries/AP_RangeFinder/AP_RangeFinder_LightWareI2C.cpp]: ohh... this is one of the weird devices, without registers. We need a new method in Device:

```c++
bool read(uint8_t *recv, uint32_t recv_len)
{
    return transfer(nullptr, 0, recv, len);
}
```

**Comment by lucasdemarchi** [comment]
See some comments above. And double check the coding style, making sure unrelated changes don't sneak in.

File: libraries/AP_RangeFinder/RangeFinder.cpp (hunk 2 above)
Comment (lgywata) [inline libraries/AP_RangeFinder/RangeFinder.cpp]: Ack.

File: libraries/AP_RangeFinder/AP_RangeFinder_LightWareI2C.h (hunk 3 above)
Comment (lgywata) [inline libraries/AP_RangeFinder/AP_RangeFinder_LightWareI2C.h]: Ack.

File: libraries/AP_RangeFinder/AP_RangeFinder_LightWareI2C.cpp (hunk 4 above)
Comment (lgywata) [inline libraries/AP_RangeFinder/AP_RangeFinder_LightWareI2C.cpp]: ok, thanks.

**Comment by lgywata** [comment]
see #3910
